File: arraysAndHashing/encode-and-decode-strings/java/OptimalSolution.java
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;

public class OptimalSolution {

    public String encode(List<String> strs){
        StringBuilder result = new StringBuilder();
        for(String s: strs){
            result.append(s.length()).append('#').append(s);
        }
        return result.toString();
    }

    public List<String> decode(String str){

        List<String> result = new ArrayList<>();
        int i =0;
        while(i < str.length()){
            int j=i;
            while(str.charAt(j)!='#'){
                j++;
            }
            int length = Integer.parseInt(str.substring(i,j));
            i=j+1;
            j=i+length;
            result.add(str.substring(i,j));
            i=j;
        }
        return result;
    }

    public static void main(String[] args){
        String[] input = new String[] {"neet","code","love","you"};
        String[] input1 = new String[] {"we","say",":","yes"};
        List<String> decodedList = new ArrayList<>();
        OptimalSolution solution = new OptimalSolution();
        List<String> inputList = Arrays.asList(input);
        List<String> inputList1 = Arrays.asList(input1);
        String encodedString = solution.encode(inputList);
        System.out.println("Encoded String: "+encodedString);
        decodedList = solution.decode(encodedString);
        System.out.println("Decoded List: "+decodedList);
        decodedList.clear();
        String encodedString1 = solution.encode(inputList1);
        System.out.println("Encoded String: "+encodedString1);
        decodedList = solution.decode(encodedString1);
        System.out.println("Decoded List: "+decodedList);
    }
}

File: arraysAndHashing/encode-and-decode-strings/csharp/OptimalSolution.cs
using System;
using System.Collections.Generic;
using System.Text;

 public class OptimalSolution{
    public static void Main(string[] args){
        string[] input = new string[] {"neet","code","love","you"};
        string[] input1 = new string[] {"we","say",":","yes"};
        List<string> decodedList = new List<string>();
        OptimalSolution solution = new OptimalSolution();
        List<string> inputList = new List<string>(input);
        List<string> inputList1 = new List<string>(input1);
        string encodedString = solution.encode(inputList);
        Console.WriteLine("Encoded String: " + encodedString);
        decodedList = solution.decode(encodedString);
        Console.WriteLine("Decoded List: " + string.Join(", ", decodedList));
        decodedList.Clear();
        string encodedString1 = solution.encode(inputList1);
        Console.WriteLine("Encoded String: " + encodedString1);
        decodedList = solution.decode(encodedString1);
        Console.WriteLine("Decoded List: " + string.Join(", ", decodedList));
    }

    public string encode(List<string> strs){
        StringBuilder result = new StringBuilder();
        foreach(string s in strs){
            result.Append(s.Length).Append('#').Append(s);
        }
        return result.ToString();
    }

    public List<string> decode(string str){
        List<string> result = new List<string>();
        int i = 0;  
        while(i < str.Length){
            int j = str.IndexOf('#', i);
            int length = int.Parse(str.Substring(i, j - i));
            result.Add(str.Substring(j + 1, length));
            i = j + 1 + length;
        }
        return result;
    }
 }

